File: week13/보연/PG_GPS.cpp
#include <vector>
#include <cstring>
#include <algorithm>

using namespace std;

const int SIZE = 201;

int solution(int n, int m, vector<vector<int>> edge_list, int k, vector<int> gps_log) {
	int answer = 0;
	vector<int> edge[SIZE];
	//int dp[k][n+1];
	//memset(dp, 2e8, sizeof(dp));
	vector<vector<int>> dp;
	dp.assign(k, vector<int>(n + 1, 2e9));
	for (int i = 0; i < edge_list.size(); i++) {
		edge[edge_list[i][0]].push_back(edge_list[i][1]);
		edge[edge_list[i][1]].push_back(edge_list[i][0]);
	}

	dp[0][gps_log[0]] = 0;
	for (int i = 1; i < k; i++) { //시간
		for (int j = 1; j <= n; j++) { //거점 번호
			dp[i][j] = min(dp[i][j], dp[i - 1][j] + (gps_log[i] == j ? 0 : 1));
			for (int d = 0; d < edge[j].size(); d++) { //인접한 노드
				int next = edge[j][d];
				dp[i][j] = min(dp[i][j], dp[i - 1][next] + (gps_log[i] == j ? 0 : 1));
			}
		}
	}

	int end = gps_log[k - 1];
	if (dp[k - 1][end] == 2e9) answer = -1;
	else answer = dp[k - 1][end];
	return answer;
}

File: week13/유리/PG_GPS.java
import java.util.ArrayList;

class Solution {
    public int solution(int n, int m, int[][] edge_list, int k, int[] gps_log) {
        int [][] dp = new int [k][n+1];
		ArrayList<Integer> [] data = new ArrayList[n+1];
		for(int i = 1; i <= n; i++) {
			data[i] = new ArrayList<>();
			data[i].add(i);
		}
		
		for(int i = 0; i < m; i++) {
			int tmp = edge_list[i][0];
			int tmp2 = edge_list[i][1];
			data[tmp].add(tmp2);
			data[tmp2].add(tmp);
		}
		
		for(int i = 0; i< k; i++) {
			for(int x = 1; x <= n; x++) {
				dp[i][x] = 1001;
			}
		}
		dp[0][gps_log[0]] = 0;
		
		for(int i = 0; i <k-1; i++) {
			for(int x=1;x<=n;x++) {	//시작지점
				//System.out.println(i+"/"+x);
				if(dp[i][x] == 1001) continue;	//도달 불가
				//System.out.println("도달 가능");
				for(int y =0; y<data[x].size();y++) {
					int next = data[x].get(y);
					int nv = 0;
					if(gps_log[i+1]!=next) {
						nv = 1;
					}
					dp[i+1][next] = Math.min(dp[i+1][next], dp[i][x]+nv);
				}
			}
		}
		
		int answer = dp[k-1][gps_log[k-1]];
		if(answer == 1001) {
			answer = -1;
		}
        return answer;
    }
}
